### ISSUE TITLE
SDFGI - Massive Slowdown, When the Camera Speed is Too Fast

### ISSUE THREAD

**Issue by RaTcHeT302** [godotengine/godot#41154]
**Godot version:**
no commit hash, as it was compiled by hand, but it's probably the one starting from here
https://github.com/godotengine/godot/commit/f2149fe763a5678237b417fb9fefec5a5c74ae9d

**OS/device including version:**
windows 10 1709 16299.431
vulkan
r9 380 4 gb ggdr 5 1425 mhz

AMD stuff
2020.0729.1547.28422
20.20.01.13-200729a-357776E-RadeonSoftwareAdrenalin2020

**Issue description:**
if you increase your camera speed, to a very high value, the framerate drops into the 20s

this is bad if you were to make, a space game with a space ship, which needs to fly VERY FAST
this makes using the engine impossible, the framerate is too low and it's barely usable at all times

**Steps to reproduce:**
1. select the file system menu
2. double click on default_env.res
3. in the inspector menu, enable sdfgi
4. select the view port
5. move the camera around and increase the camera's movement speed until it starts to stutter (scroll wheel up)
6. your framerate is destroyed
![](https://i.imgur.com/bzcUBVp.png)

**Minimal reproduction project:**
even an empty project will do

**Comment by Calinou** [comment]
Does it improve if you go to the Project Settings, decrease the number of SDFGI rays to 5 (the lowest)and increase the number of frames to convert to 30 (the slowest)? These are the settings I used in https://github.com/godotengine/godot/pull/39965.

**Comment by RaTcHeT302** [comment]
i already lowered the number of cascades and it did nothing at all

it just looks worse and the framerate is still really low
![](https://i.imgur.com/wClMYXV.png)

the probe ray count only goes up to 8 not 5
![](https://i.imgur.com/NCJtFQS.png)

changing the settings didn't help and the framerate is still low, it's beyond unusable
![](https://i.imgur.com/aOOy5yB.png), 
![](https://i.imgur.com/P0stagL.png)

**Comment by Zylann** [comment]
In such scenarios I would imagine SDFGI should blend towards a much lower resolution, larger-scale area to adjust for a higher-speed motion scenario, which is typically matching what happens with a spaceship. It goes faster, but further away from the map than a human character would be.

**Comment by RaTcHeT302** [comment]
> 
> 
> In such scenarios I would imagine SDFGI should blend towards a much lower resolution, larger-scale area to adjust for a higher-speed motion scenario (or be turned off?), which is typically matching what happens with a spaceship. It goes faster, but further away from the map than a human character would be.

my framerate shouldn't drop into the 20s when i have an empty scene, all i did was to increase the camera's speed

**Comment by Zylann** [comment]
It will drop at least because it tries to relocate SDF volumes at a much faster rate. Something in there is probably not dependent on scene complexity.
Another thing that could occur is to limit how many times such "relocation" occurs, but @reduz may know more about this.

**Comment by RaTcHeT302** [comment]
here's an even more barebones scene to showcase the problem
![](https://i.imgur.com/On07uqR.png)

**Comment by RaTcHeT302** [comment]
> If you think it should not happen _specifically within an empty scene_, then it's easily fixed by having SDFGI not run when the scene is empty, but that's probably not the fix you are expecting.

???

it should never do this in the first place, like, you have a scene with 3D models, the framerate drops, you have an empty scene, the framerate drops 

there's clearly a problem in here, saying that, i should just never use sdfgi, seems like nonsense to me, i mean, to me it just makes the whole technique pointless, if it's not usable in 99% of most games

i mean immagine a racing game where you framerate drops this badly, do you just make all the cars, go really slowly? that's ridiculous

if someone were to make a wipeout styled game, they wouldn't be able to ever take advantage of the lighting system

i guess you can make a walking simulator with this system, but i don't know, i just find it odd how nobody ever brought this up, seems really strange to me

**Comment by Calinou** [comment]
@RaTcHeT302 The `master` branch is a perpetual work-in-progress. You have to expect new features to not always work as expected :slightly_smiling_face:

**Comment by RaTcHeT302** [comment]
> @RaTcHeT302 The `master` branch is a perpetual work-in-progress. You have to expect new features to not always work as expected 🙂

yes i know, but i still felt like this was very important to bring up, this seems like a pretty critical problem to me, and i'm surprised that nobody pointed it out yet, seems very odd to me

**Comment by RaTcHeT302** [comment]
> There are cascades but maybe they could bve dynamic themselves? i.e if the game uses 4 cascades at once, upon scripted decision it could fade to 4 higher cascades (rather than being more than 4 all the time to sort of support them all, if you see what I mean)
> 
> ```
> Walking       Flying         Area size
> 
> ---           Cascade 3      64
> ---           Cascade 2      32
> ---           Cascade 1      16
> Cascade 3     Cascade 0      8
> Cascade 2     ---            4
> Cascade 1     ---            2
> Cascade 0     ---            1
> ```

guys it's not about the flying specifically, it's about the player's speed in itself, i just gave a very general example

but immagine if your guy started to sprint, or if you had like, a crysis style turbo speed sprint, your framerate would pretty much be obliterated

or take just cause 3 as an example, your character can use a slingshot, to attach himself to objects very far away
once you attach yourself to a building, you are very quickly dragged towards it
![](https://i.imgur.com/KzCz7hO.png)

now as your character moves across the map, SDFGI would need to update all those lights, and your framerate would just suffer as a whole, it would be unplayable by any normal means

i'm surprised that this was never taken into account at all, point is that, if you have a character in a space ship, on foot, whatever, with a speed that's higher beyond walking speed, your framerate will suffer as a result, and it makes the system really not versatile at all, and it makes it's use case, extremelly limited unless this is fixed

**Comment by RaTcHeT302** [comment]
but see the fact that this happens with a scene that's just plain empty confuses me the most, i mean there are no lights in there, it was literally pitch black darkness, so what's going on here?

**Comment by RaTcHeT302** [comment]
okay i'll update the title, maybe that'll clear things up, just in case

**Comment by reduz** [comment]
This is expected, cascades need to re-voxel when you move, if you make a game that will move fast, you will need to use a larger cell size so revoxeling happens lest often. That said there is a good amount of optimization pending for SDFGI that will happen when everything else is more or less done, so will see this again later.

**Comment by reduz** [comment]
Oh also, the cmvalley scene you are testing with specifically sucks for SDFGI if you move fast, because it consists of very few extremely complex objects (in the million of polygons), so even if you want to revoxel a small bit of the scene when you move, its going to send millions of polygons to the vertex shader. A Properly made game scene that has the map split into more sections and objects will work way better.

**Comment by 31** [comment]
> please stop fixating on the flying thing, and understand that my framerate DROPPED when there was NOTHING IN THE SCENE

Maybe something very specific that's useful to bring up here: an empty scene has as much *volume* as a complex scene, it just happens to be empty. Since SDFGI (with my limited understanding) maintains a decent amount of volume-based data, I would expect it to have some framerate impact even on an empty scene.

(On the flipside, if there are optimizations that help a lot with empty space specifically, then I imagine they should help with the empty space in complex scenes too--so on first glance I'm not sure empty scene perf needs special attention.)

**Comment by reduz** [comment]
Also as a note, jump flood after revoxel will run anyway even if your scene is completely empty, this will drop framerate a bit if it happens to often, so make sure to use a larger cell size. 

All these things will be present in the documentation by the time its written, so for the time being understand that you are testing a work in progress, undocumented feature.
